### PR TITLE
Implement classes for nodes and process management.

### PR DESCRIPTION
**Pull request by csadorf**
Add classes:

 - nodes.NodesTreeWidget
 - process.ProcessMonitor
 - process.NodesTreeWidget

All new classes are exposed in the root name space.

### CODE REVIEW

**Review by yakutovicha — COMMENTED** [review]
Thanks, @csadorf.

As we discussed already, it would be nice if you could integrate the `ProcessMonitor` and the  `ProcessFollowerWidget`.

Then I could directly put it to the process view notebook and test it there.

**Comment by csadorf** [comment]
@yakutovicha I believe there might be some other widgets that should use the `ProcessMonitor`, like `ProcessList`. Should we adjust that in the same PR or different PR?

**Comment by CasperWA** [comment]
> @yakutovicha I believe there might be some other widgets that should use the `ProcessMonitor`, like `ProcessList`. Should we adjust that in the same PR or different PR?

I vote for different PR to at least starting getting this stuff in and tweak as we go?

**Comment by csadorf** [comment]
@yakutovicha This is ready for final review and merge. Could you please have a look at this?

File: aiidalab_widgets_base/process.py [hunk 1]
```diff
@@ -1,16 +1,26 @@
...
 from inspect import isclass
+from inspect import signature
+from threading import Event
+from threading import Lock
+from threading import Thread
 from time import sleep
-import warnings
-import pandas as pd
+
+# External imports
```
Comment (yakutovicha) [inline aiidalab_widgets_base/process.py:13]:
```suggestion
# External imports.
```

**Review by yakutovicha — COMMENTED** [review]
thanks, @csadorf, very nice work. Just a couple of small comments.

File: aiidalab_widgets_base/process.py [hunk 2]
```diff
@@ -1,16 +1,26 @@
 """Widgets to work with processes."""
 # pylint: disable=no-self-use
-
+# Built-in imports
```
[the remark this answers is not in the document]
Comment (csadorf) [reply aiidalab_widgets_base/process.py:3]: This is not a sentence? 😕

File: aiidalab_widgets_base/process.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (csadorf) [reply aiidalab_widgets_base/process.py:13]: see above

File: aiidalab_widgets_base/process.py [hunk 3]
```diff
@@ -700,3 +708,111 @@ def start_autoupdate(self, update_interval=10):
...
+
+    Args:
+        refresh_period:
+            The time period in between updates to the process tree view in seconds.
+    """
+
+    process = traitlets.Instance(ProcessNode, allow_none=True)
+    selected_nodes = traitlets.Tuple(read_only=True).tag(trait=traitlets.Instance(Node))
+
+    def __init__(self, refresh_period=0.2, **kwargs):
```
[the remark this answers is not in the document]
Comment (csadorf) [reply aiidalab_widgets_base/process.py:789]: This should not be part of this class. An attribute can very easily be attached to the instance or if you really need to have a class, just wrap it:
```
class ProcessNodesTreeWidgetWithTitle(ProcessNodesTreeWidget):
    title = "Process Tree"
```

File: aiidalab_widgets_base/process.py [hunk 4]
```diff
@@ -700,3 +708,111 @@ def start_autoupdate(self, update_interval=10):
...
+        self._tree = NodesTreeWidget()
+        self._tree.observe(self._observe_tree_selected_nodes, ["selected_nodes"])
+
+        if refresh_period > 0:
```
[the remark this answers is not in the document]
Comment (csadorf) [reply aiidalab_widgets_base/process.py:793]: Fair enough.

**Comment by csadorf** [comment]
@yakutovicha Can we merge this?

File: aiidalab_widgets_base/process.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/process.py]: I am a little bit confused here. Why the widget that belongs to the widgets-base can't just be directly used in the `ProcessFollower` without additional modifications?

It is fine with me to discuss whether we want to make `ProcessFollower` obsolete. However, we should do it in a separate issue/PR where we then remove all the `title` attributes from the other process-related classes.

File: aiidalab_widgets_base/process.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/process.py:3]: I am not a language expert, but I believe a sentence can even consist of one word. My suggestion, though, was more about styling. I just want to keep the same comment styling that I used everywhere in this package:
```python
# Comment.
```

[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/process.py:3]: Anyways, this is not a hard requirement. Up to you if you want to apply this styling to your comments.

**Comment by yakutovicha** [comment]
> @yakutovicha Can we merge this?

I think we first need to agree on the `title` attribute. I put my reasoning there.

File: aiidalab_widgets_base/process.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yakutovicha) [reply aiidalab_widgets_base/process.py]: For the record: the `title` attribute has been introduced in https://github.com/aiidalab/aiidalab-widgets-base/pull/90 following the same principles put in the `StructureManager` API (https://github.com/aiidalab/aiidalab-widgets-base/pull/94) originally proposed in https://github.com/aiidalab/aiidalab-widgets-base/issues/65.

[the remark this answers is not in the document]
Comment (csadorf) [reply aiidalab_widgets_base/process.py]: Ok, sounds good!

**Comment by csadorf** [comment]
> > @yakutovicha Can we merge this?
> 
> I think we first need to agree on the `title` attribute. I put my reasoning there.

Right, I've added the attribute.

**Review by yakutovicha — APPROVED** [review]
great work, thanks a lot @csadorf !

Please merge once the tests are passed.

**Pull request merged**